tools/build_langserver/langserver/reference.go

```diff
@@ -0,0 +1,105 @@
+package langserver
+
+import (
+	"context"
+	"core"
+	"encoding/json"
+	"fmt"
+	"path/filepath"
+	"plz"
+	"query"
+
+	"tools/build_langserver/lsp"
+
+	"github.com/sourcegraph/jsonrpc2"
+)
+
+const referenceMethod = "textDocument/references"
+
+func (h *LsHandler) handleReferences(ctx context.Context, req *jsonrpc2.Request) (result interface{}, err error) {
+	if req.Params == nil {
+		return nil, nil
+	}
+
+	var params lsp.ReferenceParams
+	if err := json.Unmarshal(*req.Params, &params); err != nil {
+		return nil, err
+	}
+
+	documentURI, err := getURIAndHandleErrors(params.TextDocument.URI, referenceMethod)
+	if err != nil {
+		return nil, err
+	}
+
+	h.mu.Lock()
+	defer h.mu.Unlock()
+
+	refs, err := h.getReferences(ctx, documentURI, params.Position)
+	if err != nil && len(refs) == 0 {
+		log.Warning("error occurred when trying to get references: %s", err)
+		return nil, nil
+	}
+
+	return refs, nil
+}
+
+func (h *LsHandler) getReferences(ctx context.Context, uri lsp.DocumentURI, pos lsp.Position) ([]lsp.Location, error) {
+	def, err := h.analyzer.BuildDefsFromPos(ctx, uri, pos)
+	if def == nil || err != nil {
+		return nil, err
+	}
+
+	label, err := getCoreBuildLabel(def, uri)
+	if err != nil {
+		return nil, err
+	}
+
+	//Ensure we do not get locked out
+	h.analyzer.State.NeedBuild = false
```

Comment:
I'm not sure you should persist the state object on the analyzer, that may cause some weird results if it gets reused concurrently.

Why not create a new one each time?